src/features/orders/Timeline.tsx

```diff
@@ -0,0 +1,246 @@
+import React, { useCallback, useEffect } from 'react'
+import {
+    Timeline as TimelineComponent,
+    StepProps,
+} from '../../share/timeline'
+import LocalShippingIcon from '@material-ui/icons/LocalShipping'
+import SendIcon from '@material-ui/icons/Send'
+import Icon from '@material-ui/core/Icon'
+import DoneIcon from '@material-ui/icons/Done'
+import { useSelector, useDispatch } from 'react-redux'
+import { selectors, actions } from './reducer'
+import { makeStyles, Theme, useTheme } from '@material-ui/core/styles'
+import DetailLongTextWrapper from './DetailLongTextWrapper'
+import Box from '@material-ui/core/Box'
+import { MAP_STATUS_ID_TO_TEXT } from './constants'
+import RejectedIcon from '@material-ui/icons/Close'
+import { types } from './reducer'
+
+const DISABLE_COLOR_OFFSET = 500;
+
+const useDoneStyles = makeStyles({
+    icon: {
+        color: '#0cb500',
+    },
+});
+
+export default function Timeline() {
+    const dispatch = useDispatch();
+    const theme = useTheme();
+
+    const activeStep = useSelector(selectors.timeline.activeStep);
+    const orderId = useSelector(selectors.detail.id);
+    const order = useSelector(selectors.ordersTable.data)[orderId];
+    const status = order.order_status;
+
+    const handleStepChange = useCallback(
+        (id: number) => {
+            dispatch(actions.timeline.changeActiveStep(id));
+        },
+        [dispatch],
+    );
+
+    useEffect(() => {
+        dispatch(
+            actions.timeline.changeActiveStep(
+                MAP_ORDER_STATUS_TO_STEP_NAME[status],
+            ),
+        );
+    }, [status, dispatch]);
+
+    return (
+        <TimelineComponent
+            steps={generateSteps(order, theme)}
+            activeStep={
+                activeStep < 0 ?
+                    MAP_ORDER_STATUS_TO_STEP_NAME[status] :
+                    activeStep
+            }
+            onStepChange={handleStepChange}
+        ></TimelineComponent>
+    );
+}
+
+enum StepName {
+    Ordered = 0,
+    Processing,
+    Complete,
+    Cancel,
+}
+
+interface MapOrderStatusToStepName {
+    [index: number]: number;
+}
+
+const MAP_ORDER_STATUS_TO_STEP_NAME: MapOrderStatusToStepName = {
+    [types.OrderStatus.Ordered]: StepName.Ordered,
+    [types.OrderStatus.Processing]: StepName.Processing,
+    [types.OrderStatus.Complete]: StepName.Complete,
+    [types.OrderStatus.Cancel]: StepName.Cancel,
+}
+
+interface MapStepNameToOrderStatus {
+    [index: number]: number;
+}
+
+const MAP_STEP_NAME_TO_ORDER_STATUS: MapStepNameToOrderStatus = {
+    [StepName.Ordered]: types.OrderStatus.Ordered,
+    [StepName.Processing]: types.OrderStatus.Processing,
+    [StepName.Complete]: types.OrderStatus.Complete,
+    [StepName.Cancel]: types.OrderStatus.Cancel,
+}
+
+function generateSteps(
```

Comment:
it might make sense to extract this logic into separate file because now this file looks a bit overloaded